daffodil-runtime1/src/main/scala/org/apache/daffodil/infoset/SAXInfosetOutputter.scala

```diff
@@ -0,0 +1,303 @@
+    }
+
+    contentHandler.startElement(ns, elemName, qName, attrs)
+  }
+
+  private def doEndElement (diElem: DIElement, contentHandler: ContentHandler): Unit = {
+    val (ns: String, elemName: String, qName: String) = getNameSpaceElemNameAndQName(diElem)
+    contentHandler.endElement(ns, elemName, qName)
+    doEndPrefixMapping(diElem, contentHandler)
+  }
+
+  private def getNameSpaceElemNameAndQName(
+    diElem: DIElement): (String, String, String) = {
+    val ns: String =
+      if (diElem.erd.namedQName.namespace.isNoNamespace) {
+        ""
+      } else {
+        diElem.erd.namedQName.namespace.toString
+      }
+    val elemName = diElem.erd.namedQName.local
+    val qName = diElem.erd.namedQName.toQNameString
+    (ns, elemName, qName)
+  }
+
+}
+
+class DaffodilOutputContentHandler(out: OutputStream, pretty: Boolean = false)
+  extends ContentHandler with Indentable with XMLInfosetOutputter {
+  private val writer = new OutputStreamWriter(out)
+  private var prefixMapping: NamespaceBinding = null
+  private val outputNewlineStack: MStackOfBoolean = {
+    val s = MStackOfBoolean()
+    s.push(false)
+    s
+  }
+
+  // if the top of the stack is true, we have guessed we should output a newline
+  private def outputNewline: Boolean = !outputNewlineStack.isEmpty && outputNewlineStack.top
+
+  def reset(): Unit = {
+    resetIndentation()
+    writer.flush()
+    prefixMapping = null
+    outputNewlineStack.clear()
+    out.flush()
+  }
+
+  override def setDocumentLocator(locator: Locator): Unit = {
+    // do nothing
+  }
+
+  override def startDocument(): Unit = {
+    writer.write("<?xml version=\"1.0\" encoding=\"UTF-8\" ?>")
+  }
+
+  override def endDocument(): Unit = {
+    writer.write(System.lineSeparator())
+    writer.flush()
+  }
+
+  override def startPrefixMapping(prefix: String, uri: String): Unit = {
+    val _prefix = if (prefix == "") null else prefix
+    prefixMapping = NamespaceBinding(_prefix, uri, prefixMapping)
+  }
+
+  override def endPrefixMapping(prefix: String): Unit = {
+  // do nothing
+  }
+
+  override def startElement(uri: String, localName: String, qName: String, atts: Attributes): Unit = {
+    // the pop/true removes whatever is on the stack which is our previous guess for whether we
+    // would need a newline after the previous end tag. As we are currently at the start of a new
+    // tag, we want to correct that assumption (in case it was false)
+    outputNewlineStack.pop()
+    outputNewlineStack.push(true)
+    if (pretty) {
+      writer.write(System.lineSeparator())
+      outputIndentation(writer)
+    }
+    // handle start of tag
+    writer.write("<")
+    writer.write(qName)
+    // handle attributes
+    for(i <- 0 until atts.getLength) {
+      val attsValue = atts.getValue(i)
+      val attsQName = atts.getQName(i)
+      writer.write(s""" $attsQName="$attsValue"""")
+    }
+    // handle namespaces
+    if(prefixMapping != null) {
+      val pm = prefixMapping.toString()
```

Comment:
Add space after ``if` and ``for`` above. Double check the if statements, there's a couple more in this file I think.